controllers/awscluster_controller.go

```diff
@@ -66,48 +66,83 @@ var (
+	elbServiceFactory     awsClusterELBServiceFactory
+	securityGroupFactory  awsClusterSecurityGroupServiceFactory
 	Endpoints             []scope.ServiceEndpoint
 	WatchFilterValue      string
 }
 
-// getEC2Service factory func is added for testing purpose so that we can inject mocked EC2Service to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getEC2Service(scope scope.EC2Scope) services.EC2Interface {
-	if r.ec2ServiceFactory != nil {
-		return r.ec2ServiceFactory(scope)
+type NewClusterReconcilerInput struct {
+	Manager          ctrl.Manager
+	WatchFilterValue string
+	Endpoints        []scope.ServiceEndpoint
+}
+
+type awsClusterEC2ServiceFactory func(scope.EC2Scope) services.EC2Interface
+
+func withAWSClusterEC2ServiceFactory(fn awsClusterEC2ServiceFactory) awsClusterReconcilerOption {
+	return func(r *awsClusterReconciler) {
+		r.ec2ServiceFactory = fn
 	}
-	return ec2.NewService(scope)
 }
 
-// getELBService factory func is added for testing purpose so that we can inject mocked ELBService to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getELBService(scope scope.ELBScope) services.ELBInterface {
-	if r.elbServiceFactory != nil {
-		return r.elbServiceFactory(scope)
+type awsClusterELBServiceFactory func(scope.ELBScope) services.ELBInterface
+
+func withAWSClusterELBServiceFactory(fn awsClusterELBServiceFactory) awsClusterReconcilerOption {
+	return func(r *awsClusterReconciler) {
+		r.elbServiceFactory = fn
 	}
-	return elb.NewService(scope)
 }
 
-// getNetworkService factory func is added for testing purpose so that we can inject mocked NetworkService to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getNetworkService(scope scope.ClusterScope) services.NetworkInterface {
-	if r.networkServiceFactory != nil {
-		return r.networkServiceFactory(scope)
+type awsClusterNetworkServiceFactory func(scope.ClusterScope) services.NetworkInterface
+
+func withAWSClusterNetworkServiceFactory(fn awsClusterNetworkServiceFactory) awsClusterReconcilerOption {
+	return func(r *awsClusterReconciler) {
+		r.networkServiceFactory = fn
 	}
-	return network.NewService(&scope)
 }
 
-// getSecurityGroupService factory func is added for testing purpose so that we can inject mocked SecurityGroupService to the AWSClusterReconciler.
-func (r *AWSClusterReconciler) getSecurityGroupService(scope scope.ClusterScope) services.SecurityGroupInterface {
-	if r.securityGroupFactory != nil {
-		return r.securityGroupFactory(scope)
+type awsClusterSecurityGroupServiceFactory func(scope.ClusterScope) services.SecurityGroupInterface
+
+func withAWSClusterSecurityGroupServiceFactory(fn awsClusterSecurityGroupServiceFactory) awsClusterReconcilerOption {
+	return func(r *awsClusterReconciler) {
+		r.securityGroupFactory = fn
 	}
-	return securitygroup.NewService(&scope, awsSecurityGroupRoles)
+}
+
+func NewClusterReconciler(input NewClusterReconcilerInput, opt ...awsClusterReconcilerOption) *awsClusterReconciler {
```

Comment:
We could probably get this to return an interface that `awsClusterReconciler` implements? As it stands we are returning an non-exported type.

Reply:
@richardcase The problem with using `reconcile.Reconciler` as a return type here, would be, we won't be able to call SetUpWithManager() here as it doesn't exists for `reconcile.Reconciler` but it does for `awsClusterReconciler`

Reply:
and only awsClusterReconciler will implement that Reconciler? or awsMachineReconciler as well?

Reply:
I think they all could...all the controllers have `Reconcile` and `SetUpWithManager` functions. wdyt?

Reply:
Sounds reasonable to me, I will give it a try. Thanks @richardcase.

Reply:
Only problem is reconcile.Normal() and reconcile.Delete() has different function signatures for `AWSMachine` and `AWSCluster`.

Reply:
And we call these func cause only creation of objects doesn't help in the tests, as `Reconcile()` func tries to create cluster scope which can't be done without mocking, and can be mocked only the way we are doing right now in tests.

Reply:
@richardcase do you have any views on this?

Reply:
Sorry @Ankitasw i lost track of this. 

I'd be interested to know why cluster scope can't be mocked, did you run into issues?

Reply:
I missed this conversation as well, let me read the code again and get back to you 🙂

Reply:
After digging a bit, found [this](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3159#discussion_r803460544) thread, which lsays why we won't be able to mock clusterScope.

Reply:
@richardcase to summarize the problem here, in all of the current tests we are checking the conditions set on Cluster through ClusterScope object, if we directly call Reconcile() func, then all those tests scenarios added would go for toss and we could have only one scenario, i.e, if reconcile is successful, so how about we leave the return type pf this func as `awsClusterReconciler` only? Or if you suggest any alternative for tests. that also would be ok.
cc @sedefsavas

Reply:
I need to have a think about this, its still not sitting well....but maybe i'm not thinking about it right :wink:

Reply:
yes please take your time 🙂 just wanted to get a conclusion so bumped this discussion.

Reply:
If we need to test for multiple conditions then we can use the fake client to get the `AWSCluster` and then look at the conditions.

@Ankitasw -  I will pull your branch down and have a play around with the code....if you are ok with that?

Reply:
@richardcase would you be able to look at this sometime this week, wanted to conclude this discussion.

Reply:
It will probably have to be next week as today is my last working day this week.

Reply:
@richardcase any luck with this?